### ISSUE TITLE
Cannot prevent LSP servers from starting automatically

### ISSUE THREAD

**Issue by dradetsky** [doomemacs/doomemacs#8116]
I would like to enable LSP support for various modes, but without necessary starting the LSP server automatically when I visit a file whose mode has LSP support enabled. The reason is that I often want to just refer to a single file in a large project while working on another large project, and if this automatically starts an LSP instance then pretty soon I have dozens of of them running in huge projects and emacs starts having issues.

It was suggested that the correct approach for this issue was to modify the hooks. The following example was given:

```
;; i want flycheck, but only on-demand. read ~/.emacs.d/lisp/doom.el for a list
;; of hooks and their order of execution
(remove-hook 'doom-first-buffer-hook #'global-flycheck-mode)

;; i want lsp, but only on-demand.
(remove-hook! '(c-mode-local-vars-hook
               c++-mode-local-vars-hook
               objc-mode-local-vars-hook
               cmake-mode-local-vars-hook)
  #'lsp!)
```

While I confirmed that this does in fact work for c-mode, I cannot seem to make any variation of it work for csharp-mode no matter what I do. In particular:

```
(remove-hook! '(csharp-mode-local-vars-hook
                csharp-tree-sitter-mode-local-vars-hook
                c-mode-local-vars-hook
                c++-mode-local-vars-hook
                objc-mode-local-vars-hook
                cmake-mode-local-vars-hook)
  #'lsp!)
```

Does not work; editing a csharp file will still spawn an omnisharp instance. For reference, here is a repo with a user doomdir which replicates the issue for csharp mode: https://github.com/dradetsky/doom-replicate-lsp-thing.  In addition, I have just confirmed that the issue affects the python module as well.

Looking at the cc module, we see that LSP is enabled via the following:

```
(when (modulep! +lsp)
  (add-hook! '(c-mode-local-vars-hook
               c++-mode-local-vars-hook
               objc-mode-local-vars-hook
               cmake-mode-local-vars-hook
               cuda-mode-local-vars-hook)
             :append #'lsp!)
```

whereas in the csharp module we have

```
(use-package! csharp-mode
  :hook (csharp-mode . rainbow-delimiters-mode)
...
  (when (modulep! +lsp)
    (add-hook 'csharp-mode-local-vars-hook #'lsp! 'append))
```

and in the python module

```
(use-package! python
  :mode ("[./]flake8\\'" . conf-mode)
...
  (when (modulep! +lsp)
    (add-hook 'python-mode-local-vars-hook #'lsp! 'append)
    ;; Use "mspyls" in eglot if in PATH
    (when (executable-find "Microsoft.Python.LanguageServer")
      (set-eglot-client! 'python-mode '("Microsoft.Python.LanguageServer"))))
```

In other words, in both cases the operation to add the `#'lsp` hook occurs inside the `use-package!` form. I would assume that the solution would be to use `after!` as in something like

```
      (after! python
        (remove-hook! python-mode-local-vars-hook #'lsp))
```

However, I can't seem to make this work either.

**Comment by dradetsky** [comment]
Also: I looked into the possibility of wrapping the body of `#'lsp` in a check which would exit immediately unless it was being run interactively, or if the user had enabled a flag `+lsp-start-automatically`. I haven't PR'd this yet on the grounds that I wasn't sure if I was doing something wrong.

**Comment by hlissner** [comment]
In order for your `remove-hook` calls to work, you need to execute them after the `lsp!` function has been added to their respective `*-mode-local-vars-hook` hook variable, but that'd require a slew of `(after! PACKAGE ...)` blocks, which will end up being quite a lot of work for such a small change. To save you trouble, I'd suggest you simply advise the `lsp!` function. This will stop it from activating automatically across the board:

```elisp
;;; add to $DOOMDIR/config.el
(advice-add #'lsp! :override #'ignore)
```

This will do as you want without affecting `M-x lsp-mode` (to activate it on demand).

As this isn't a bug and our issue tracker isn't for general user support, I will close this, but feel free to follow up here.
